### PR TITLE
Convert a Spotify URL to a Youtube Video URL

### PR DESCRIPTION
**Pull request by Nomad145**
This PR adds functionality to Tardibot that will allow users not subscribed to Spotify to listen to Spotify links as Youtube Videos.

### CODE REVIEW

File: app/BotMan/ConfigParser.php [hunk 1]
```diff
@@ -37,10 +45,59 @@ public function configure(BotMan $bot)
...
+     *
+     *     App\Conversation\MyConversation@handle
+     *
+     * @param string $conversation The conversation string.
+     * @return string $conversation The conversation string.
+     */
+    private function addConversation(string $conversation) : string
+    {
+        if (!$this->isConversationValid($conversation)) {
+            throw new \UnexpectedValueException(sprintf(
```
Comment (michaelwpearson) [inline app/BotMan/ConfigParser.php]: Is there a reason you didn't import `UnexpectedValueException`?

File: app/BotMan/ConfigParser.php [hunk 2]
```diff
@@ -37,10 +45,59 @@ public function configure(BotMan $bot)
...
+     *
+     * @param string $reply The conversation string.
+     * @return true If the conversation is formatted correctly and both the
+     *              conversation/method exist.
+     */
+    private function isConversationValid(string $reply) : bool
+    {
+        [$class, $method] = explode('@', $reply);
+
+        if (null == $class || null == $method) {
```
Comment (michaelwpearson) [inline app/BotMan/ConfigParser.php]: I recommend using the `===` for `null` checks.

File: app/BotMan/ConfigParser.php [hunk 3]
```diff
@@ -37,10 +45,59 @@ public function configure(BotMan $bot)
      *
      * @param string $reply
      */
-    private function reply(string $reply)
+    private function addReply(string $reply) : Callable
     {
         return function (BotMan $bot) use ($reply) {
```
Comment (michaelwpearson) [inline app/BotMan/ConfigParser.php]: How does `$bot` get injected into the `addReply()` scope?

File: app/BotMan/Conversation/SpotifyUrlConversation.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace App\BotMan\Conversation;
+
+use BotMan\BotMan\Messages\Conversations\Conversation;
+use BotMan\BotMan\BotMan;
+use App\SpotifyTrackConverter;
+use App\Spotify\Track;
```
Comment (michaelwpearson) [inline app/BotMan/Conversation/SpotifyUrlConversation.php]: Nitpick: I like to alphabetize my imports. It's super easy with PHPStorm, and I'm sure there's a key combo (maybe via a plugin) for VIM.

File: app/Spotify/Track.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * @author Michael Phillips <michael.phillips@realpage.com>
+ */
+class Track
+{
+    private const SPOTIFY_URL = 'https://open.spotify.com/track';
+
+    /**
+     * @param string $id The Spotify Track ID.
+     */
+    private function __construct(string $id)
```
Comment (michaelwpearson) [inline app/Spotify/Track.php]: Interesting design choice to make the constructor private on a VO. I like it.

File: app/Spotify/Track.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Validate the URL.
+     *
+     * @param string $url
+     * @throws UnexpectedValueException If $url is not a Spotify Track URL.
+     * @return void
+     */
+    private static function validateUrl(string $url) : void
```
Comment (michaelwpearson) [inline app/Spotify/Track.php]: Typically, I put all validation rules in my VO's constructor, so they automatically run when I call `return new self()`.

File: app/Youtube/Video.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace App\Youtube;
+
+/**
+ * @author Michael Phillips <michael.phillips@realpage.com>
+ */
+class Video
```
Comment (michaelwpearson) [inline app/Youtube/Video.php]: Suggestion: I would name this class more descriptively as something like `VideoUrl`.

File: tests/Integration/SpotifyTrackConverterTest.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        $this->subject = new SpotifyTrackConverter($this->youtube, $this->spotify);
+    }
+
+    public function testConvert()
+    {
+        $track = Track::fromUrl('https://open.spotify.com/track/37G9ACbVFCdZvdHVSA3dxz?si=iZxqHIsiR7W_2dK6hUWanQ');
+
+        $video = $this->subject->convert($track);
+
+        $this->assertEquals('https://youtube.com/watch?v=Tu9KgGqXDyw', $video->getUrl());
```
Comment (michaelwpearson) [inline tests/Integration/SpotifyTrackConverterTest.php]: I see you tested the "happy path". What about the "sad path"?

File: app/BotMan/ConfigParser.php (hunk 3 above)
Comment (Nomad145) [reply app/BotMan/ConfigParser.php]: `addReply` returns a callable accepting the `$bot` argument.

```
$func = $parser->addReply('foo');

$func($bot);
```

File: app/BotMan/ConfigParser.php (hunk 1 above)
Comment (Nomad145) [reply app/BotMan/ConfigParser.php]: I typically don't import classes in the root namespace.  I personally don't see a point, but I'm not firm on this point.

File: app/BotMan/Conversation/SpotifyUrlConversation.php (hunk 4 above)
Comment (Nomad145) [reply app/BotMan/Conversation/SpotifyUrlConversation.php]: I started doing this after reading this review.  I like it a lot, actually.

File: app/Spotify/Track.php (hunk 6 above)
Comment (Nomad145) [reply app/Spotify/Track.php]: Since my constructor is private, this method is called from the `fromUrl` factory method.  See line 27.

**Pull request merged**